app/cypress/integration/cif/project-revision/undo-changes.spec.js

```diff
@@ -0,0 +1,179 @@
+    cy.get('[label*="Due Date"]').should("have.length", 0);
+
+    // undo annual reports
+    cy.findByText(/Annual reports/i).click();
+    cy.findByText(/Add annual reports/i).click();
+    cy.addAnnualReport(1, "2000-05-05", "2000-07-23");
+    cy.findByRole("button", { name: /undo changes/i }).click();
+    cy.findByText(/Annual Report 1/i).should("not.exist");
+
+    // undo milestone reports
+    cy.findByText(/Milestone reports/i).click();
+    cy.findByText(/Add milestone reports/i).click();
+    cy.addMilestoneReport(
+      1,
+      "I am a description",
+      "General",
+      "1990-08-12",
+      "1991-04-17",
+      "Professional Engineer"
+    );
+    cy.findByRole("button", { name: /undo changes/i }).click();
+    cy.findByText(/Milestone Report 1/i).should("not.exist");
+    cy.get('[label*="Due Date"]').should("have.length", 0);
+  });
+
+  it("undoes changes on an existing project when the user clicks the Undo Changes button", () => {
+    cy.sqlFixture("dev/004_cif_project");
+    cy.sqlFixture("dev/005_cif_reporting_requirement");
+    cy.sqlFixture("dev/006_commit_project_revision");
+    cy.mockLogin("cif_admin");
+    cy.visit("/cif/projects");
+
+    cy.findAllByRole("button", { name: /view/i }).first().click();
+    cy.findByRole("button", { name: /edit/i }).click();
+
+    //undo overview
+    cy.findByLabelText(/Proposal Reference/i)
+      .clear()
+      .type("I will be undone");
+    cy.findByLabelText(/Proposal Reference/i).should(
+      "have.value",
+      "I will be undone"
+    );
+    cy.findByRole("button", { name: /undo changes/i }).click();
+    cy.contains("Changes saved.");
+
+    cy.findByLabelText(/Proposal Reference/i).should("have.value", "001");
+
+    // undo managers
+    cy.findByText(/Project Details/i).click();
+    cy.findByText(/Edit project managers/i).click();
+    cy.findByLabelText(/tech team primary/i).click();
+    cy.contains(/Swanson, Ron/).click();
+    cy.findByLabelText(/tech team primary/i).should(
+      "have.value",
+      "Swanson, Ron"
+    );
+    cy.findByRole("button", { name: /undo changes/i }).click();
+    cy.contains("Changes saved.");
+    cy.findByLabelText(/tech team primary/i).should(
+      "have.value",
+      "Testuser, cif_internal"
+    );
+
+    // undo contacts
+    // TODO: add this test back in once undoing contacts bug is fixed
+    // cy.findByText(/Edit project contacts/i).click();
+    // cy.findByLabelText(/Primary contact/i).click();
+    // cy.contains(/Loblaw006, Bob006/).click();
+    // cy.findByLabelText(/Primary contact/i).should(
+    //   "have.value",
+    //   "Loblaw006, Bob006"
+    // );
+    // cy.findByRole("button", { name: /undo changes/i }).click();
+    // cy.findByRole("button", { name: /undo changes/i }).click();
+    // cy.contains("Changes saved.");
+    // cy.wait(2000);
+    // cy.findByLabelText(/Primary contact/i).should("have.value", "");
+
+    // undo quarterly reports
```

Comment:
To shorten the test, could we could test one type of report?

Reply:
Definitely, since the `UndoFormChanges` is its own component. We'll just need to make sure it's covered for each form in a jest test, that the component is being rendered with the right `formChangeIds`